src/ConstructionBase.jl

```diff
@@ -24,129 +24,145 @@ for (name, path) in [
     end
 end
 
-@generated function constructorof(::Type{T}) where T
-    getfield(parentmodule(T), nameof(T))
+@static if isdefined(Base, Symbol("@assume_effects"))
+    using Base: @assume_effects
+else
+    macro assume_effects(_, ex)
+        Base.@pure ex
+    end
 end
+struct PropertyNames{syms} end
 
-constructorof(::Type{<:Tuple}) = tuple
-constructorof(::Type{<:NamedTuple{names}}) where names =
-    NamedTupleConstructor{names}()
+struct PropertyIndices{inds} end
 
-struct NamedTupleConstructor{names} end
+const PropertyKeys{K} = Union{PropertyNames{K},PropertyIndices{K}}
```

Comment:
Why are `PropertyNames` and `PropertyIndices` two distinct structs? Also, the user is only supposed to overload `PropertyNames` is that correct?

Reply:
So we can tell that one contains symbols and the other contains integers. It was just a small convenience for writing code but I can merge them.

Reply:
I think it is fine to have this distinction internally as long as the user only needs to overload `PropertyNames`.